[view-transitions] Occasional fade while animating on https://codepen.io/argyleink/pen/dyLNgpX.

mattwoodrow opened the pull request:
#### 61f529e65a3fce776e8c04a65ffda08901656751
<pre>
[view-transitions] Occasional fade while animating on <a href="https://codepen.io/argyleink/pen/dyLNgpX.">https://codepen.io/argyleink/pen/dyLNgpX.</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272184">https://bugs.webkit.org/show_bug.cgi?id=272184</a>
&lt;<a href="https://rdar.apple.com/125931538">rdar://125931538</a>&gt;

Reviewed by Tim Nguyen.

Captured view-transition snapshots need to be rescaled when the intrinsic size of the image is
different to the layout size of the pseduo element.

The captured image covers the overflow rect of the original element, but the border-box subset
defines the intrinsic size.
This computes the scale factors required to scale the intrinsic size of the up the replaced
content size of the pseudo renderer. It then scales the overflow rect (in coordinates of
the original element) into coordinates for the pseudo renderer.

Painting of the pseudo element (both new and old variants) are updated to factor this
scale in to the draw commands or layer attributes.

Doing this sizing during painting correctly exposed bugs where the computed layout of
the pseduo renderer was incorrect.

ViewTransition:::copyElementBaseProperties is changed to use the &apos;frameRect&apos; of the view
(instead of the &apos;contentsWidth&apos;) when computing the size of the root element, so that
space taken up by scrollbars isn&apos;t removed.

RenderObject&apos;s calculation of the &apos;IsAnonymous&apos; flag is changed so that the view-transition
pseudo elements are not considered anonymous and then correctly can be containing blocks
for their descendants.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateChildrenTransformAndAnchorPoint):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::shouldBeAnonymous):
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionContainer const):
(WebCore::RenderObject::hasIntrinsicAspectRatio const):
(WebCore::RenderObject::node const):
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::paintReplaced):
(WebCore::RenderViewTransitionCapture::layout):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
(WebCore::RenderViewTransitionCapture::scale const):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::createRendererIfNeeded):

Canonical link: <a href="https://commits.webkit.org/277636@main">https://commits.webkit.org/277636@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a00eea1c6614710ab7c8050e88e5cf135f8da96a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27263 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50743 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24755 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39298 "Exiting early after 10 failures. 96 tests run. 1 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48639 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24932 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41563 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20396 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42738 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6111 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44395 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43174 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52642 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23105 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19456 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46617 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; layout-tests (exception)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24369 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41722 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45472 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10627 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25171 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -622,22 +619,24 @@ Ref<MutableStyleProperties> ViewTransition::copyElementBaseProperties(Element& e
-        props->setProperty(CSSPropertyWidth, CSSPrimitiveValue::create(size.width(), CSSUnitType::CSS_PX));
-        props->setProperty(CSSPropertyHeight, CSSPrimitiveValue::create(size.height(), CSSUnitType::CSS_PX));
+        CheckedPtr renderBox = dynamicDowncast<RenderBoxModelObject>(element.renderer());
+        if (renderBox)
```
```suggestion
        if (CheckedPtr renderBox = dynamicDowncast<RenderBoxModelObject>(element.renderer()))
```

nt1m commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -241,6 +241,7 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr<RenderOb
         IsFragmentedFlow = 1 << 1,
         IsTextControl = 1 << 2,
         IsSVGBlock = 1 << 3,
+        IsViewTransitionContainer = 1 << 4,
```
I don't quite understand how this flag is applied, I only see it applied on ::view-transition and ::view-transition-new, I'm not sure I understand the logic there.

mattwoodrow replied:
It should be applied on `::view-transition`, `::view-transition-group` and `::view-transition-image-pair`. All the view transition pseudo elements that are containers.

`::view-transition-new` and `::view-transition-old` should be the ones that don't have it.

nt1m replied:
Oh I misread the code, thanks

mattwoodrow committed:
[view-transitions] Occasional fade while animating on https://codepen.io/argyleink/pen/dyLNgpX.
https://bugs.webkit.org/show_bug.cgi?id=272184
<rdar://125931538>

Reviewed by Tim Nguyen.

Captured view-transition snapshots need to be rescaled when the intrinsic size of the image is
different to the layout size of the pseduo element.

The captured image covers the overflow rect of the original element, but the border-box subset
defines the intrinsic size.
This computes the scale factors required to scale the intrinsic size of the up the replaced
content size of the pseudo renderer. It then scales the overflow rect (in coordinates of
the original element) into coordinates for the pseudo renderer.

Painting of the pseudo element (both new and old variants) are updated to factor this
scale in to the draw commands or layer attributes.

Doing this sizing during painting correctly exposed bugs where the computed layout of
the pseduo renderer was incorrect.

ViewTransition:::copyElementBaseProperties is changed to use the 'frameRect' of the view
(instead of the 'contentsWidth') when computing the size of the root element, so that
space taken up by scrollbars isn't removed.

RenderObject's calculation of the 'IsAnonymous' flag is changed so that the view-transition
pseudo elements are not considered anonymous and then correctly can be containing blocks
for their descendants.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::copyElementBaseProperties):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateChildrenTransformAndAnchorPoint):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::shouldBeAnonymous):
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderViewTransitionContainer const):
(WebCore::RenderObject::hasIntrinsicAspectRatio const):
(WebCore::RenderObject::node const):
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::paintReplaced):
(WebCore::RenderViewTransitionCapture::layout):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
(WebCore::RenderViewTransitionCapture::scale const):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::createRendererIfNeeded):

Canonical link: https://commits.webkit.org/277636@main
